implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,616 @@
 defmodule Ockam.SecureChannel.Channel do
-  @moduledoc false
-
-  ## TODO: make that Ockam.Worker
-  use GenStateMachine, callback_mode: [:handle_event_function, :state_enter]
-
+  # For now disable this, there are several dependencies that will
+  # go away soon.
+  # credo:disable-for-this-file Credo.Check.Refactor.ModuleDependencies
+  @moduledoc """
+  Ockam Secure Channel, implementation based on AsymmetricWorker
+
+                +-----------------+                          +-----------------+
+                |                 |                          |                 |
+            +---+------+    +-----+----+                 +---+------+    +-----+----+
+  Plaintext | Address  |    |Inner     |                 |Inner     |    | Address  |  Plaintext
+  <------>  |          |    |Address   | <----[...] ---> |Address   |    |          | <------>
+            +---+------+    +-----+----+    Ciphertext   +---+------+    +-----+----+
+                |  SecureChannel  |                          |   SecureChannel |
+                +-----------------+                          +-----------------+
+
+
+  The secure channel goes throw three stages:
+    * Handshaking  (noise handshake)
+    * IdentityExchange (identity exchange and verification)
+    * Established (channel fully established and peer authenticated)
+
+  At this time, the implementation don't use a proper fsm as that's not directly supported
```

Comment:
Was the session handshake too complex to accomodate the handshake stage?

Reply:
yes found the session  code (pluggable/etc)  difficult to follow,   mainly what state was passed around,   and what were processes vs what are just callback.    Plus (right now)  handshake spand across several messages,  found it simpler to keep in a single fsm  (kind of like on the previous Channel implementation,  but not like in the Identity channel one)